src/samples/audio/detect_test.c

```diff
@@ -653,6 +675,113 @@ static int test_keyword_prepare(struct comp_dev *dev)
 	return comp_set_state(dev, COMP_TRIGGER_PREPARE);
 }
 
+
+
+uint16_t test_keyword_get_sample_valid_bytes(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	return cd->sample_valid_bytes;
+}
+
+uint32_t test_keyword_get_detected(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	return cd->detected;
+}
+
+int test_keyword_set_detected(struct comp_dev *dev, uint32_t detected)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+	int ret = 0;
+
+	cd->detected = detected;
+	return ret;
+}
+
+const int16_t *test_keyword_get_input(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	/* if input is null, the function which uses
+	 * test_keyword_get_input should return
+	 * before accessing input
+	 */
+	if (!cd->input)
+		comp_err(dev, "test_keyword_get_input() null pointer");
+
+	return cd->input;
+}
+
+int16_t test_keyword_get_input_byte(struct comp_dev *dev, uint32_t index)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	if (index < 0 || index >= IN_BUFF_SIZE * sizeof(int16_t))
+		return -EINVAL;
+	return *((unsigned char *)cd->input + index);
```

Comment:
Shouldn't `cd->input` be checked for null?

Reply:
We need to see if there is an invariant that permits that variable to be null here. Maybe it just can't (_new ensures it's not null and nothing changes it until _free comes up). Do we have such an invariant?

Reply:
@paulstelian97 @cristinailie I've looked more thorough on how `cd->input` is initialized and used, and I don't think that this can be null. @cristinailie, please correct me if I'm wrong.

In this case, @cristinailie you should not check `cd->input` for null - especially these 2 ifs, one after another, that have the same return code doesn't look too nice :) 
In cases like this you either check for null everywhere (`test_keyword_get_input_byte()`, `test_keyword_get_input_elem()`) or you don't.